files/generate_system_metrics

```diff
@@ -0,0 +1,245 @@
+    # @param [string] metric_type cpu|memory
+    # @param [string] metrics_dir The puppet_metrics_collector output directory.
+    # @param [boolean] verbose Verbose output
+    #
+    # @return [void]
+    def initialize(polling_interval, file_interval, metric_type, metrics_dir, verbose = false)
+      @polling_interval = polling_interval
+      @file_interval = file_interval
+      @metric_type = metric_type
+      @metrics_dir = metrics_dir
+      @verbose = verbose
+
+      @hostname = %x[hostname].strip
+      puts "Hostname is: #{@hostname}" if @verbose
+      FileUtils.mkdir_p(@metrics_dir) unless File.directory?(@metrics_dir)
+    end
+
+    # Run sar to collect the raw system data
+    #
+    # @author Randell Pelak
+    #
+    # @return [string] raw output from sar
+    def run_sar
+      times_to_poll = (@file_interval / @polling_interval).round
+      # sar inputs are polling interval and how many times to poll
+      comm_flags = " -r" if @metric_type =~ /memory/
+      comm = "sar #{comm_flags} #{@polling_interval} #{times_to_poll}"
+      puts "sar command is: #{comm}" if @verbose
+      %x[#{comm}]
+    end
+
+    # Gets the header fields from sar output
+    #
+    # @author Randell Pelak
+    #
+    # @param [string] sar_output
+    #
+    # @return [array] header fields
+    def get_header_fields(sar_output)
+      arr = sar_output.split("\n")
+      arr[2].split(" ")[2..-1]
+    end
+
+    # Gets the averages fields from sar output
+    #
+    # @author Randell Pelak
+    #
+    # @param [string] sar_output
+    #
+    # @return [array] average fields
+    def get_averages_fields(sar_output)
+      arr = sar_output.split("\n")
+      arr[-1].split(" ")[1..-1]
+    end
+
+    # Convert the inputted sar output into json and raise errors if the sar output is invalid
+    #
+    # @author Randell Pelak
+    #
+    # @param [string] sar_output
+    # @param [obj] time_stamp_obj Time object to use for generating the filename
+    #
+    # @raise [RuntimeError] if sar_output doesn't parse correctly
+    #
+    # @return [string] json of sar output
+    def convert_sar_output_to_json(sar_output, time_stamp_obj)
+      last_line = sar_output.split("\n")[-1]
+      puts "sar output last line:\n#{last_line}" if @verbose
+      sar_error_last_line = "sar output missing Average: on the last line:\n#{sar_output}"
+      raise(sar_error_last_line)  unless last_line =~ /^Average:/
+      header = get_header_fields(sar_output)
+      averages = get_averages_fields(sar_output)
+      sar_error_header_mismatch = "sar output headers and data columns are not equal:\n#{sar_output}"
+      raise(sar_error_header_mismatch) unless header.count == averages.count
+      hostkey = @hostname.gsub('.', '-')
+      dataset = {'time_stamp_obj' => time_stamp_obj.utc.iso8601, 'servers' => {}}
+      #combine the arrays into a hash and convert into json
+      metrics_data = Hash[header.zip(averages)]
```

Comment:
This is passing raw STDOUT string data to three methods.  It would be preferable to encapsulate the conversion of the sar string output to a defined data set.  That way if the output format of sar changes in the future only one method would need to be updated.

Suggestion:
```
# @return [hash] metrics_data
def parse_sar_output(sar_output)
  sar_data = {}
  sar_arr = sar_output.split(/\n+|\r+/).reject(&:empty?).map { |line| line.split }
  averages = sar_arr.find { |e| e.include? "Average:" }
  headers = sar_arr.find { |e| e.include? "kbmemfree" }
  metrics_data = Hash[headers.zip(averages)]
end
```

Then pass the data to `convert_metrics_to_json`.

Reply:
@johnduarte hmmm.  I played around with that.  I run into the problems getting useful error messages out.  So it got a bit longer.  But it encapsulated all the processing in one method. See what you think.